specs/PrintToPdf.md

```diff
@@ -0,0 +1,303 @@
+                Callback<ICoreWebView2PrintToPdfCompletedHandler>(
+                    [](HRESULT errorCode, LPCWSTR resultFilePath) -> HRESULT {
+                        CHECK_FAILURE(errorCode);
+                        std::wstringstream formattedMessage;
+                        formattedMessage << "ResultFilePath: " << resultFilePath;
+                        MessageBox(nullptr, formattedMessage.str().c_str(),
+                            L"Print to PDF Completed", MB_OK);
+                        return S_OK;
+                    })
+                    .Get()));
+        }
+    }
+}
+```
+
+```c#
+// Prints current page to PDF using the default path and settings. If requested,
+// the orientation is changed to landscape.
+async void PrintToPdfCmdExecuted(object target, ExecutedRoutedEventArgs e)
+{
+    CoreWebView2PrintSettings printSettings =
+        WebViewEnvironment.CreateDefaultPrintSettings();
+
+    string orientationString = e.Parameter.ToString();
+    if (printSettings != null && orientationString == "Landscape")
+    {
+        printSettings.Orientation =
+            CoreWebView2PrintOrientation.Landscape;
+    }
+    string resultFilePath = await webView.CoreWebView2.PrintToPdfAsync(
+        "" /* use default path*/, printSettings);
+    MessageBox.Show(this, resultFilePath, "Print To PDF Completed");
+}
+```
+# Remarks
+
+# API Notes
+
+
+# API Details
+```
+interface ICoreWebView2_4;
+interface ICoreWebView2Environment5;
+interface ICoreWebView2PrintSettings;
+interface ICoreWebView2PrintToPdfCompletedHandler;
+
+
+/// The orientation for printing, used by the `Orientation` property on
+/// `ICoreWebView2SPrintSettings`. Currently only printing to PDF
+/// is supported.
+[v1_enum]
+typedef enum COREWEBVIEW2_PRINT_ORIENTATION
+{
+  /// Print the page(s) in portrait orientation.
+  COREWEBVIEW2_PRINT_ORIENTATION_PORTRAIT,
+
+  /// Print the page(s) in landscape orientation.
+  COREWEBVIEW2_PRINT_ORIENTATION_LANDSCAPE,
+} COREWEBVIEW2_PRINT_ORIENTATION;
+
+[uuid(9eae81c3-fe02-4084-b475-903a4ed9252e), object, pointer_default(unique)]
+interface ICoreWebView2_4 : IUnknown {
+  /// Print the current page to PDF asynchronously with the provided settings.
+  /// See `ICoreWebView2PrintSettings` for description of settings.
+  ///
+  /// Use `resultFilePath` to specify the path to the PDF file. The host should
+  /// provide an absolute path, including file name. If the path
+  /// points to an existing file, the file will be overwritten. If
+  /// `resultFilePath` is invalid, `isSuccessful` in
+  /// `ICoreWebView2PrintToPdfCompletedHandler` is set to `FALSE` and the
+  /// returned `resultFilePath` is an empty string.
+  ///
+  /// The async `PrintToPdf` operation completes when the data has been written
+  /// to the PDF file. At this time the `ICoreWebView2PrintToPdfCompletedHandler`
+  /// is invoked.
+  ///
+  /// \snippet FileComponent.cpp PrintToPdf
+  HRESULT PrintToPdf(
```

Comment:
Just thinking about how this interacts with the rest of printing:

Is printing to PDF just one aspect of printing scenarios  (like same print scenario you're just changing if you're printing to a printer vs PDF) or is printing to PDF its own sort of scenario separate from normal printing?

Do we need to separate how printing is invoked from modifying printing settings? Script in a page can call window.print() and normally that would show the print preview dialog. Should the end dev be able to intercept that, not show print preview, programmatically set the settings, and then decide to print to pdf vs the printer?

Reply:
I think printing to PDF could be one aspect of printing like you describe. On the other hand, most customers requesting this API are coming from CEF and they have requested a similar interface to this one: https://cefsharp.github.io/api/51.0.0/html/M_CefSharp_IBrowserHost_PrintToPdf.htm

Reply:
Headless CDP has a similar method and there is active work to support it in headful mode too: https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF